lib/src/environment/state/environment_state.dart

```diff
@@ -0,0 +1,60 @@
+import 'package:flutter/widgets.dart';
+
+import 'basic_environment_state.dart';
+
+/// Values used to render a [Scene].
+/// TODO improve docs
+class EnvironmentState extends ChangeNotifier {
+  BasicEnvironmentState _basicState = BasicEnvironmentState.defaultState;
+
+  /// TODO
+  BasicEnvironmentState get basicState => _basicState;
+
+  /// TODO
+  set basicState(BasicEnvironmentState newValue) {
+    _basicState = newValue;
+    notifyListeners();
+  }
+
+  /// Stores current named environment state values.
+  Map<String, dynamic> _stateMap = <String, dynamic>{};
+
+  /// Stores default named environment state values.
+  final Map<String, dynamic> _stateMapDefaults = <String, dynamic>{};
+
+  /// Sets a default
+  void setDefault({required String key, required dynamic value}) {
+    _stateMapDefaults[key] = value;
+
+    if (!_stateMap.containsKey(key)) {
+      set(key: key, value: value);
+    }
+  }
+
+  /// Returns the value corresponding to [key] if it exists.
+  T? get<T>({required String key}) => _stateMap[key];
+
+  /// Updates the environment value corresponding to [key].
+  ///
+  /// If no default has been set for this [key], [value] will be used as the
```

Comment:
This behavior seems kind of unexpected to me! Is it kind of like the first value you set becomes the default?

Reply:
What about instead if you implement reset() so it just doesn't reset the value if there is no default set? I think that would be more predictable. Something like _stateMap.allAll(_stateMapDefaults)

Reply:
Yeah, I was a bit torn on this. The primary assumption that motivated this approach is that the first value someone uses for a key is probably what they want to reset to, and if they don't know about the `setDefault` function, they might think something is broken when the reset button is pressed and nothing happens. I agree that it's maybe not what a reasonable person looking at this API would expect though. I'll give it some extra thought.